src/EditorFeatures/Core/Implementation/FindReferences/AbstractFindReferencesService.cs

```diff
@@ -51,36 +59,125 @@ protected AbstractFindReferencesService(IEnumerable<IReferencedSymbolsPresenter>
+        /// </summary>
+        private async Task AddSymbolReferencesAsync(Document document, int position, ArrayBuilder<INavigableItem> builder, IWaitContext waitContext)
         {
-            var cancellationToken = waitContext.CancellationToken;
             var result = await this.FindReferencedSymbolsAsync(document, position, waitContext).ConfigureAwait(false);
-            if (result == null)
+            if (result != null)
             {
-                return SpecializedCollections.EmptyEnumerable<INavigableItem>();
+                var referencedSymbols = result.Item1;
+                var searchSolution = result.Item2;
+
+                var q = from r in referencedSymbols
+                        from loc in r.Locations
+                        select NavigableItemFactory.GetItemFromSymbolLocation(searchSolution, r.Definition, loc.Location);
+
+                builder.AddRange(q);
             }
+        }
 
-            var referencedSymbols = result.Item1;
-            var searchSolution = result.Item2;
+        public async Task<IEnumerable<INavigableItem>> FindReferencesAsync(Document document, int position, IWaitContext waitContext)
+        {
+            var cancellationToken = waitContext.CancellationToken;
 
-            var q = from r in referencedSymbols
-                    from loc in r.Locations
-                    select NavigableItemFactory.GetItemFromSymbolLocation(searchSolution, r.Definition, loc.Location);
+            var builder = ArrayBuilder<INavigableItem>.GetInstance();
+            await AddExternalReferencesAsync(document, position, builder, cancellationToken).ConfigureAwait(false);
+
+            // TODO: Merging references from SymbolFinder and external providers might lead to duplicate or counter-intuitive results.
+            // TODO: For now, we avoid merging and just display the results either from SymbolFinder or the external result providers but not both.
+            if (builder.Count == 0)
+            {
+                await AddSymbolReferencesAsync(document, position, builder, waitContext).ConfigureAwait(false);
+            }
 
             // realize the list here so that the consumer await'ing the result doesn't lazily cause
             // them to be created on an inappropriate thread.
-            return q.ToList();
+            return builder.ToArrayAndFree();
         }
 
         public bool TryFindReferences(Document document, int position, IWaitContext waitContext)
         {
             var cancellationToken = waitContext.CancellationToken;
+            var workspace = document.Project.Solution.Workspace;
+
+            // First see if we have any external navigable item references.
+            // If so, we display the results as navigable items.
+            var succeeded = TryFindAndDisplayNavigableItemsReferencesAsync(document, position, waitContext).WaitAndGetResult(cancellationToken);            
```

Comment:
this behavior also seems need a discussion later. current model not supporting filtering or ordering by provider seems a bit limiting. CommandHandler's pattern where provider has full control of pre/post/filtering seems more flexible especially when augmenting results.